tengine/drake.ts

```diff
@@ -0,0 +1,121 @@
+   *
+   */
+  public views_path: string = "";
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - CONSTRUCTOR /////////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Construct an object of this class.
+   *
+   * @param viewsPath - The base path to the template(s).
+   */
+
+  constructor(viewsPath: string) {
+    this.views_path = viewsPath;
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // FILE MARKER - METHODS - PUBLIC ////////////////////////////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+
+  /**
+   * Render a template file and replace all template variables with the
+   * specified data.
+   *
+   * @param template - The template to render.
+   * @param data - The data that should be rendered with the template.
+   *
+   * @remarks
+   * For example, the data could be...
+   *     ```json
+   *     {
+   *       name: "John"
+   *     }
+   *     ```
+   * and the template would render "John" in <% name %>.
+   * This data can be anything and everything. It contains the data that the
+   * template engine will use for template variable replacement.
+   * @returns The html to be rendered
+   */
+  public render(template: string, data: unknown): string {
+    let code = "with(obj) { var r=[];\n";
+    let cursor = 0;
+    let match;
+    const filepath = this.views_path + template;
+    let html: string = decoder.decode(
+      Deno.readFileSync(filepath),
+    );
+
+    // Check if the template extends another template
+    let extended = html.match(/<% extends.* %>/g);
+    if (extended) {
+      extended.forEach((m: string, i: number) => {
+        html = html.replace(m, "");
+        let template = m.replace('<% extends("', "").replace('") %>', "");
+        template = decoder.decode(
+          Deno.readFileSync(this.views_path + template),
+        );
+        html = template.replace("<% yield %>", html);
+      });
+    }
+
+    // Check for partials
+    let partials;
+    // deno-lint-ignore no-cond-assign
+    while ((partials = html.match(/<% include_partial.* %>/g))) {
+      partials.forEach((m: string, i: number) => {
+        let template = m
+          .replace('<% include_partial("', "")
+          .replace('") %>', "");
+        template = decoder.decode(
+          Deno.readFileSync(this.views_path + template),
+        );
+        html = html.replace(m, template);
+      });
+    }
+
+    // The following code was taken from (and modified):
+    // https://krasimirtsonev.com/blog/article/Javascript-template-engine-in-just-20-line
+    // Thanks, Krasimir!
+    let re = /<%(.+?)\%>/g;
+    let reExp = /(^( )?(var|if|for|else|switch|case|break|{|}|;))(.*)?/g;
+    let result;
+    function add(line: string, js: unknown | null = null) {
+      js
```

Comment:
I know that this is code from deno-drash, but I think there are a lot more ternary operators than needed. Wouldn't something like
```suggestion
    let add = (line: string, js = false) => {
      if (js) {
         code += line.match(reExp) ? line + '\n' : 'r.push(' + line + ');\n';
      } else {
         code += line != '' ? 'r.push("' + line.replace(/"/g, '\\"') + '");\n' : "";
      }
      // No need to return anything, since we're trying to write a template engine, not write a template engine in 20 lines or less
    }
```
be simpler?

Reply:
possibly. this isn't our code btw. i found it on a blog and asked the author if we could use it. i took it as is and only modified it a bit so that it could handle including partials. nothing else was done to it. i know it's messy :/

Reply:
It's not too bad though; it's still decipherable, gets the job done, and it's not [Obfuscript](https://www.gnu.org/philosophy/javascript-trap.html)

Reply:
i appreciate the feedback 🙂